autofix: also run dry-run tests

hannesm opened the pull request:
it unfolds as follows: there was a test_autofix_osemfail and a test_autofix - the former had `@pytest.mark.parametrize("dryrun", [True, False], ids=["dryrun", "not-dryrun"])`, the latter `@pytest.mark.parametrize("dryrun", [False], ids=["not-dryrun"])`

Some time later, we managed to remove the osemfail test definition (since osemgrep passes all these tests) -- and now with this PR, we enable the dry-run tests again (and also make them pass on osemgrep by doing some additional work when converting the matches / when to emit "fixed_lines").

Result for `make osempass` is now 217 passes (and 127 failures) --> 63%

hannesm committed:
osemgrep: track applied fixes

hannesm committed:
re-enable dry-run autofix tests

aryx commented on src/osemgrep/reporting/Cli_json_output.ml:
```diff
@@ -345,19 +345,41 @@ let cli_match_of_core_match ~dryrun (hrules : Rule.hrules) (m : OutJ.core_match)
-              | [], _
-              | _, [] ->
-                  [])
+          Option.fold ~none:None
```
probably better to move this code in a separate helper function.
Also I'm not a fan of those Option.fold, Option.value ~default ...
in general I find the code that does explciit 'match x with Some y -> ... | None -> ...' clearer that those Option.fold.

aryx replied:
Option.iter is fine (better than Option.fold ~note:()), 
let* is fine.
But I often find Option.fold and Option.value making the code more complicated.

hannesm replied:
done in [a943e44](https://github.com/semgrep/semgrep/pull/9648/commits/a943e44754bd4c3cf6bcd5ef96c612d877603677)

aryx commented on src/osemgrep/reporting/Cli_json_output.mli:
```diff
@@ -14,6 +14,7 @@ val exit_code_of_error_type : Semgrep_output_v1_t.error_type -> Exit_code.t
 (* internals used also for incremental display of matches *)
 val cli_match_of_core_match :
   dryrun:bool ->
+  ?applied_fixes:(string, (int * int) list) Hashtbl.t ->
```
So this hash is actually populated by the function? it's an OUT parameter?
Would it be possible to return it instead with cli_match?
At least add a comment that this parameter is an OUT parameter modifed by the function
and used in the caller to do ... ?

hannesm replied:
I documented the purpose:
```
   The [applied_fixes] hash table is both in and out parameter, it is used for
   deciding whether a fixed_lines elements is included in the cli_match. This
   depends on whether an overlapping fix was already included in an earlier
   cli_match in the same list of matches.
```

hope that makes it clear.

aryx reviewed (changes requested):
Request changes for adding comments on the hash OUT parameter and the rewriting of the
helper.

hannesm committed:
address review comments:
- document the ?applied_fixes argument
- extract make_fixed_lines into a helper function

hannesm committed:
reword
